src/contracts/AccessListFactory.ts

```diff
@@ -0,0 +1,144 @@
+import { BigNumber, Signer } from 'ethers'
+import { Config } from '../config'
+import AccessListFactory from '@oceanprotocol/contracts/artifacts/contracts/accesslists/AccessListFactory.sol/AccessListFactory.json'
+import { generateDtName, sendTx, getEventFromTx, ZERO_ADDRESS } from '../utils'
+import { AbiItem, ReceiptOrEstimate } from '../@types'
+import { SmartContractWithAddress } from './SmartContractWithAddress'
+
+/**
+ * Provides an interface for Access List Factory contract
+ */
+export class AccesslistFactory extends SmartContractWithAddress {
+  getDefaultAbi() {
+    return AccessListFactory.abi as AbiItem[]
+  }
+
+  /**
+   * Instantiate AccessListFactory class
+   * @param {string} address The factory contract address.
+   * @param {Signer} signer The signer object.
+   * @param {string | number} [network] Network id or name
+   * @param {Config} [config] The configuration object.
+   * @param {AbiItem[]} [abi] ABI array of the smart contract
+   */
+  constructor(
+    address: string,
+    signer: Signer,
+    network?: string | number,
+    config?: Config,
+    abi?: AbiItem[]
+  ) {
+    super(address, signer, network, config, abi)
+    this.abi = abi || this.getDefaultAbi()
+  }
+
+  /**
+   * Create new Access List Contract
+   * @param {AccessListData} listData The data needed to create an NFT.
+   * @param {Boolean} [estimateGas] if True, return gas estimate
+   * @return {Promise<string|BigNumber>} The transaction hash or the gas estimate.
+   */
+  public async deployAccessListContract<G extends boolean = false>(
+    nameAccessList: string,
+    symbolAccessList: string,
+    tokenURI: string[],
+    transferable: boolean = false,
+    owner: string,
+    user: string[],
+    estimateGas?: G
+  ): Promise<G extends false ? string : BigNumber> {
+    if (!nameAccessList || !symbolAccessList) {
+      const { name, symbol } = generateDtName()
+      nameAccessList = name
+      symbolAccessList = symbol
+    }
+    const estGas = await this.contract.estimateGas.deployAccessListContract(
+      nameAccessList,
+      symbolAccessList,
+      transferable,
+      owner,
+      user,
+      tokenURI
+    )
+    if (estimateGas) return <G extends false ? string : BigNumber>estGas
+    // Invoke createToken function of the contract
+    const tx = await sendTx(
+      estGas,
+      this.signer,
+      this.config?.gasFeeMultiplier,
+      this.contract.deployAccessListContract,
+      nameAccessList,
+      symbolAccessList,
+      transferable,
+      owner,
+      user,
+      tokenURI
+    )
+    const trxReceipt = await tx.wait()
+    const events = getEventFromTx(trxReceipt, 'NewAccessList')
```

Comment:
what if the tx fails, need to provide a proper response for the client otherwise the line bellow  events.args[0] will fail with uncaught error

Reply:
Yes, the same should be applied in NFTFactory when creating a NFT, try-catch block is missing from there too.

Reply:
created https://github.com/oceanprotocol/ocean.js/pull/1853

Reply:
yes but the logic here still need few more changes, what if the tx fails, there would be no events matching the name we pass, so for at line 89 you will end up with an "cannot read property x of undefined"

Reply:
I put a check only for `tx` returned by `sendTx`, because if that is not present the code below will fail.

Reply:
i dont't really get it maybe i'm missing something, so if sendTx fails it will only log an error cause here we don't have a try catch or something handling that, so it go on wi calling `tx.wait()`, then trying to access events for that tx etc

Reply:
if `sendTx` will fail, either will throw an error that will be caught in the try-catch block within `sendTx` function or the try-catch block from `deployAccessListContract` function which is surrounded by, either the tx receipt is null. If tx returned from `sendTx` does not exist, this will throw an error with message: `Tx for deploying new access list was not processed on chain.`. It will not go to `tx.wait()` please check my latest code where I have added the condition, because it is above `tx.wait()`.